src/components/application_manager/src/application_manager_impl.cc

```diff
@@ -3224,15 +3224,17 @@ void ApplicationManagerImpl::UnregisterApplication(
   ApplicationSharedPtr app_to_remove;
   connection_handler::DeviceHandle handle = 0;
   {
```

Comment:
Without these changes, sometimes a deadlock occurs. In future we'll prepare fix taht will remove direct usages of applications_. It will be used only through its data accessor to solve the problem.
**DataAccessor<ApplicationSet> ApplicationManagerImpl::applications() const {
  DataAccessor<ApplicationSet> accessor(applications_, applications_list_lock_ptr_);
  return accessor;
}**

Reply:
Sorry, it's difficult to show the exact changes I'm suggesting, but notice that there are two scope blocks here:
```
  { // first
    { // second
      sync_primitives::AutoLock lock(applications_list_lock_ptr_);
      auto it_app = applications_.begin();
      while (applications_.end() != it_app) {
        if (app_id == (*it_app)->app_id()) {
          app_to_remove = *it_app;
          applications_.erase(it_app++);
          break;
        } else {
          ++it_app;
        }
      }
    } // second
    ...
  } // first, line 3298
```
What I'm suggesting is that we reduce the scope of the "first" block rather than adding a new block entirely, since there is no need for two scope blocks here, both are intended to cover `sync_primitives::AutoLock lock(applications_list_lock_ptr_);`.
```
  {
    sync_primitives::AutoLock lock(applications_list_lock_ptr_);
    auto it_app = applications_.begin();
    while (applications_.end() != it_app) {
      if (app_id == (*it_app)->app_id()) {
        app_to_remove = *it_app;
        applications_.erase(it_app++);
        break;
      } else {
        ++it_app;
      }
    }
+ }
  ...
- } // line 3298
```

Reply:
Here is a test commit on my fork which shows the exact change I'm suggesting: https://github.com/jacobkeeler/sdl_core/commit/ba72cd076be9578ac7e61546ee80eb4e0c550aa5

Reply:
Oh, now I get your point. Fixed with https://github.com/smartdevicelink/sdl_core/pull/3617/commits/2454a6483bd866a3166be685a3814cce8ecbb497